nodeport: fix CT accounting in revDNAT path

julianwiedmann opened the pull request:
While working on https://github.com/cilium/cilium/issues/22659, I noticed that we mis-account RX traffic in the nodeport revDNAT path. Background is that we first do a CT lookup for a packet (and update the statistics while at it), but then decide that the matched CT entry doesn't fit the criteria for nodeport revDNAT.

```release-note
Fix double-accounted RX packets in CT statistics when Nodeport is in use.
```

julianwiedmann committed:
bpf: nodeport: drop reply by local backend if revDNAT is skipped

RevDNAT for replies from a local service backend is handled by tail-calling
from bpf_lxc to CILIUM_CALL_IPV*_NODEPORT_REVNAT with
bpf_skip_recirculation() set.

If rev_nodeport_lb*() then doesn't find a matching CT entry, don't return
CTX_ACT_REDIRECT to the caller. Without a CT entry we also didn't perform
a FIB lookup, so `ifindex` is still 0 and the subsequent bpf_redirect()
won't do any good.

As bpf_lxc only performs the tail-call if ct_state->node_port is set,
finding no related Nodeport CT entry is unexpected. So drop the packet.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: short-cut the Ingress path if no revDNAT is needed

The nodeport Ingress path for reply traffic (ie. everything that is *not*
a service request) first applies RevSNAT. This targets:
- BPF-masqueraded connections (IPv4 only)
- EgressGW reply traffic (IPv4 only)
- host-originating connections
- reply traffic by remote backends

We only need to continue in the RevDNAT code to handle replies by remote
backends[1]. So in a pure DSR config, we can recircle straight back to the
program start.

[1] and to redirect EgressGW reply traffic that piggy-backs on the
    RevDNAT's tunnel code.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

YutaroHayakawa reviewed (approved):
LGTM, thanks!

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2005,6 +2006,18 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __u32 *ifind
+
+	if (!ct_has_nodeport_egress_entry4(get_ct_map4(&tuple), &tuple))
+		goto skip_nat;
+#endif /* ENABLE_MASQUERADE */
```
Thinking we might actually need to match the full condition in https://github.com/cilium/cilium/blob/4ddeb34a9192a3634d816a915c12629cac5340b7/bpf/bpf_host.c#L1196_L1201. And thus also add a similar check in the IPv6 path.

And bpf_overlay is even more [inclusive](https://github.com/cilium/cilium/blob/4ddeb34a9192a3634d816a915c12629cac5340b7/bpf/bpf_overlay.c#L563) in its SNAT usage :thinking: .

julianwiedmann replied:
> Thinking we might actually need to match the full condition in https://github.com/cilium/cilium/blob/4ddeb34a9192a3634d816a915c12629cac5340b7/bpf/bpf_host.c#L1196_L1201. And thus also add a similar check in the IPv6 path.
> 
Yeah, just checked. The CT entries for hostns-originating connections (which we track due to potential SNAT conflicts vs forwarded Nodeport connections) get the same RX double accounting. Will update.

julianwiedmann committed:
bpf: remove check for ENABLE_EGRESS_GATEWAY in to-netdev's SNAT path

EgressGW depends on BPF-Masquerading, and without ENABLE_MASQUERADE none
of the EgressGW-specific SNAT logic will even be compiled in. So this check
is redundant and can go away.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -1110,15 +1110,16 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
 
-	return CTX_ACT_REDIRECT;
+	if (bpf_skip_recirculation(ctx))
+		return DROP_NAT_NO_MAPPING;
```
Makes sense to drop given the ifindex 0.. so there's no known call-site where this was set?

julianwiedmann replied:
> Makes sense to drop given the ifindex 0.. so there's no known call-site where this was set?

Not sure I follow. The `from-container` path sets the `skip-recirculation` flag before it tail-calls, and we **should** always find a matching CT entry. So this just handles the "impossible" condition that we don't find a CT entry.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1091,15 +1091,16 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
 
-	return CTX_ACT_REDIRECT;
+	if (bpf_skip_recirculation(ctx))
+		return DROP_NAT_NO_MAPPING;
```
If I understand correctly the impact of this commit, in case of a mismatch between `egress_ct_state->node_port` and `ingress_ct_state` (i.e., first is true but second doesn't exist for some reason), we would silently drop the packet because we would call `bpf_redirect` with an incorrect ifindex. Now, we will emit a Cilium drop event.

So clearly a good change. And not a bugfix because we don't expect this mismatch to ever happen today (correct me if wrong @borkmann).

Nevertheless, it is a change in behavior, so could we explicitly mention it in the commit message? That is, mention the user-visible impact.

julianwiedmann replied:
> Nevertheless, it is a change in behavior, so could we explicitly mention it in the commit message?

At this stage I don't expect it to have any user-visible impact (ie. it's totally an "impossible to hit" condition). But we do need it in the following patches to have a clean `goto out` path.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1640,14 +1644,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
-	 * needed)
+	 * CALL_IPV4_FROM_NETDEV in the code above.
+	 */
+#if !defined(ENABLE_DSR) || (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)) ||	\
```
I think it would be worth having a code comment here to mention why we have a condition on DSR all of a sudden. IIUC, something like "Reverse DNAT for DSR was performed at the backend's node so we can skip that case for the intermediate node here."

julianwiedmann replied:
Added a comment

> If we're not in full DSR mode, reply traffic from remote backends might pass back through the LB node and requires revDNAT.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1640,14 +1644,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
+	 * CALL_IPV4_FROM_NETDEV in the code above.
+	 */
+#if !defined(ENABLE_DSR) || (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)) ||	\
+    (defined(ENABLE_EGRESS_GATEWAY) && !defined(TUNNEL_MODE))
```
Could you expand on this case in a code comment or in the commit? It seems like the trickiest case, yet it's relegated to a footnote :sweat_smile:  
In particular, why does `TUNNEL_MODE` make a difference here? Doesn't egress gateway traffic go through the tunnel regardless of the routing mode?

julianwiedmann replied:
Just to be clear - you saw the comments right below this line, but they're not sufficient?

pchaigno replied:
Yes I did. The only comment below on the egress gateway case describes what is being done, not why:

> Also redirect EgressGW reply traffic into tunnel.

julianwiedmann replied:
Ah - yes, we're piggy-backing on the RevDNAT code for EgressGW replies. The details are in https://github.com/cilium/cilium/blob/1b4f2a334a8f3826c925a5bf49b7fc1165f292f4/bpf/lib/nodeport.h#L1976-L1981

Yes, that's not pretty - and we should probably do a s/CILIUM_CALL_IPV4_NODEPORT_REVNAT/CILIUM_CALL_IPV4_NODEPORT_NAT_INGRESS_CONTINUE to clarify that it's also handling some non-RevDNAT cases.

pchaigno replied:
Ouch! I see. We probably need a code comment here as well; it may be enough to point to "the egress gateway special case in rev_nodeport_lb{4,6}".

julianwiedmann replied:
Added a comment:

> Also let rev_nodeport_lb4() redirect EgressGW reply traffic into tunnel (see there for details).

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1995,11 +1992,12 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __u32 *ifind
+	if (ret == DROP_CT_UNKNOWN_PROTO)
+		goto out;
+	if (IS_ERR(ret))
+		return ret;
```
What's the rationale for this change? If it's just a performance optimization, note that it likely increases verifier complexity; I'm not sure the tradeoff is worth it.

Do we expect any user impact or is it just supposed to bypass useless code? The `IS_ERR(ret)` case in particular makes me think there might be some user impact.

julianwiedmann replied:
We shouldn't try to extract the L4 ports from a packet with unknown / unsupported L4 protocol.

pchaigno replied:
But we already check for the unknown protocol cases (i.e., not UDP, TCP, SCTP, or ICMP) in `ct_lookup{4,6}`. So why do we need to duplicate the check here?

julianwiedmann replied:
Ah, I see what you mean. In patch 4 we could/should tolerate the `DROP_CT_UNKNOWN_PROTO` and just continue. But for patch 5 we then strictly need that check - `ct_has_nodeport_egress_entry4()` expects the full 5-tuple, so we need to manually load the ports before even calling into `ct_lookup()`.

[long-term we can add an optimized ct_lookup_with_tuple() that doesn't extract the ports a second time. I suspect that would apply in other places too]

pchaigno replied:
Ah, I see. That makes sense. Could we have a small code comment to explain that?

julianwiedmann replied:
I ended up squashing those two patches together, and using a new LB helper that handles the port extraction internally.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -992,21 +996,23 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
+		return DROP_CT_INVALID_HDR;
+
+	if (!ct_has_nodeport_egress_entry6(get_ct_map6(&tuple), &tuple))
+		goto out;
```
IIUC, only this last commit is a bugfix, so only this commit should be backported to stable branches. Unless I'm missing some dependency between the commits (?), the four previous commits can go into another pull request and don't need to be backported.

julianwiedmann replied:
We could arguably defer patches 2/3. But I believe they're worth having in this series to clarify the current behavior, and self-document what sort of traffic is actually affected by the subsequent changes in the RevDNAT code.

ie. if we missed any condition in the "this needs to continue in `CILIUM_CALL_IPV4_NODEPORT_REVNAT`" check, that's an indication that the RevDNAT changes themselves might also be incomplete.

pchaigno reviewed (changes requested):
Tricky stuff. Thanks for looking into this logic :muscle: It's definitely needed.

Couple comments below. Most are to make sure I understood and ask for more details in commits/code comments. The last one is to discuss what we should actually backport.

julianwiedmann committed:
bpf: nodeport: skip CT lookup for non-NodePort replies in RevDNAT path

Consider a pod-to-external connection that gets masqueraded in to-netdev.
A reply for such a connection first gets revSNATed in
tail_nodeport_nat_ingress_ipv4(), and then passes into rev_nodeport_lb4()
to check whether it's actually a reply to a forwarded NodePort connection.

Here the CT lookup succeeds, but then the additional checks for
ct_state.node_port and ct_state.rev_nat_index fail. Therefore the reply
tail-calls back into CILIUM_CALL_IPV4_FROM_NETDEV and eventually gets
delivered towards the pod. On the pod's ingress path there is a second
(expected) CT lookup.

As the first CT lookup succeeded, we end up double-accounting the reply
in __ct_lookup().

A similar scenario affects host-to-external connections. With non-DSR
NodePort they are also in scope of our SNAT code (to avoid clashes with
forwarded Nodeport connections), and get connection-tracked. The
external-to-host reply traffic will then do a CT lookup in the revSNAT
path, and the gratuitious CT lookup in the revDNAT path results in
double accounting.

So before calling ct_lookup4() in the RevDNAT nodeport path, first check if
we will find a relevant CT entry. As ct_has_nodeport_egress_entry4()
expects a full 5-tuple, add a new LB-specific helper that extracts the
tuple. Make the helper generic enough (ie. handle ICMP) so that we can
re-use it for the SVC lookup path in the future.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
